storages/backends/gcloud.py

```diff
@@ -0,0 +1,233 @@
+        if not file_object.blob:
+            raise IOError(u'File does not exist: %s' % name)
+        return file_object
+
+    def _save(self, name, content):
+        cleaned_name = clean_name(name)
+        name = self._normalize_name(cleaned_name)
+
+        content.name = cleaned_name
+        encoded_name = self._encode_name(name)
+        file = GoogleCloudFile(encoded_name, 'rw', self)
+        file.blob.upload_from_file(content, size=content.size)
+        return cleaned_name
+
+    def delete(self, name):
+        name = self._normalize_name(clean_name(name))
+        self.bucket.delete_blob(self._encode_name(name))
+
+    def exists(self, name):
+        if not name:  # root element aka the bucket
+            try:
+                self.bucket
+                return True
+            except ImproperlyConfigured:
+                return False
+
+        name = self._normalize_name(clean_name(name))
+        return bool(self.bucket.get_blob(self._encode_name(name)))
+
+    def listdir(self, name):
+        name = self._normalize_name(clean_name(name))
+        # for the bucket.list and logic below name needs to end in /
+        # But for the root path "" we leave it as an empty string
+        if name and not name.endswith('/'):
+            name += '/'
+
+        files_list = list(self.bucket.list_blobs(prefix=self._encode_name(name)))
+        files = []
+        dirs = set()
+
+        base_parts = name.split("/")[:-1]
+        for item in files_list:
+            parts = item.name.split("/")
+            parts = parts[len(base_parts):]
+            if len(parts) == 1 and parts[0]:
+                # File
+                files.append(parts[0])
+            elif len(parts) > 1 and parts[0]:
+                # Directory
+                dirs.add(parts[0])
+        return list(dirs), files
+
+    def _get_blob(self, name):
+        # Wrap google.cloud.storage's blob to raise if the file doesn't exist
+        blob = self.bucket.get_blob(name)
+
+        if blob is None:
+            raise NotFound(u'File does not exist: {}'.format(name))
+
+        return blob
+
+    def size(self, name):
+        name = self._normalize_name(clean_name(name))
+        blob = self._get_blob(self._encode_name(name))
+        return blob.size
+
+    def modified_time(self, name):
+        name = self._normalize_name(clean_name(name))
+        blob = self._get_blob(self._encode_name(name))
+        return timezone.make_naive(blob.updated)
+
+    def get_modified_time(self, name):
+        name = self._normalize_name(clean_name(name))
+        blob = self._get_blob(self._encode_name(name))
+        updated = blob.updated
+        return updated if setting('USE_TZ') else timezone.make_naive(updated)
+
+    def url(self, name):
+        # Preserve the trailing slash after normalizing the path.
+        name = self._normalize_name(clean_name(name))
+        blob = self._get_blob(self._encode_name(name))
```

Comment:
Yes we could, but there are strange cases like the one you highlighted above. The way things are written now, it's easy to verify that `gcloud` works in the same way as `s3boto`.

I can make that change in this PR if you really want me to, but I think the right way to do it would be to merge this PR first then clean up this backend, `s3boto`, and any other backend that does this sort of thing in a separate PR.